File: Apps/iOS/Features/Harmonica/Touch/SquareTouchMapper.swift
import CoreGraphics
import HarmonicaCore

struct SquareTouchMapper {
    private let touches: [FingerTouch]
    private let size: CGSize

    // MARK: - Public

    init(_ touches: [FingerTouch], across size: CGSize) {
        self.touches = touches
        self.size = size
    }

    var pitchShaping: PitchShaping? {
        leadingLocation.map { PitchShaping(clamping: 1 - 2 * Double($0.y / size.height)) }
    }

    var vibrato: VibratoDepth? {
        leadingLocation.map { VibratoDepth(clamping: Double($0.x / size.width)) }
    }

    var marks: [FingerMark] {
        let leading = leadingLocation

        return touches.indices.map { index in
            FingerMark(
                id: index,
                location: touches[index].location,
                diameter: FingerMark.restingDiameter,
                isTheDecidingFinger: touches[index].location == leading
            )
        }
    }

    // MARK: - Private

    private var leadingLocation: CGPoint? {
        touches.map(\.location).min { $0.y < $1.y }
    }
}

File: Web/src/features/harmonica/touch/squareTouchMapper.ts
import type { AreaSize } from './areaSize.js'
import type { Contact } from './contact.js'
import type { FingerMark } from './fingerMark.js'
import { restingDiameter } from './fingerMark.js'

export interface SquareShaping {
    readonly pitch: number
    readonly vibrato: number
}

export class SquareTouchMapper {
    constructor(
        private readonly contacts: readonly Contact[],
        private readonly size: AreaSize
    ) {}

    shaping(): SquareShaping | null {
        const leading = this.leading()
        if (leading === undefined || this.isAPinch()) return null

        return {
            pitch: 1 - (2 * leading.y) / this.size.height,
            vibrato: leading.x / this.size.width
        }
    }

    marks(): FingerMark[] {
        const leading = this.leading()
        return this.contacts.map(contact => ({
            x: contact.x,
            y: contact.y,
            diameter: restingDiameter,
            isTheDecidingFinger: contact === leading
        }))
    }

    private isAPinch(): boolean {
        return this.contacts.length > 1
    }

    private leading(): Contact | undefined {
        let topmost: Contact | undefined
        for (const contact of this.contacts) {
            if (topmost === undefined || contact.y < topmost.y) topmost = contact
        }
        return topmost
    }
}
